test/trigger/test_trigger.js

```diff
@@ -390,6 +405,136 @@ let testBadZoneId = function(file, contents) {
+        if (!funcStr.includes(builtInResponseStr))
+          continue;
+        const output = { responseOutputStrings: undefined };
+        // Call the function to get the outputStrings.
+        response = trigger.response({}, {}, output);
+        outputStrings = output.responseOutputStrings;
+
+        if (typeof outputStrings !== 'object') {
+          errorFunc(`${file}: '${trigger.id}' built-in response did not set outputStrings.`);
+          continue;
+        }
+      } else {
+        if (trigger.outputStrings && typeof outputStrings !== 'object') {
+          errorFunc(`${file}: '${trigger.id}' outputStrings must be an object.`);
+          continue;
+        }
+        outputStrings = trigger.outputStrings;
+      }
+
+      // TODO: should we prevent `output['phrase with spaces']()` style constructions?
+      // TODO: should we restrict outputStrings keys to valid variable characters?
+
+      // TODO: share this with popup-text.js?
+      const paramRegex = /\${\s*([^}\s]+)\s*}/g;
+
+      // key => [] of params
+      const outputStringsParams = {};
+
+      // For each outputString, find and validate all of the parameters.
+      for (const key in outputStrings) {
+        let templateObj = outputStrings[key];
+        if (typeof templateObj !== 'object') {
+          errorFunc(`${file}: '${key}' in '${trigger.id}' outputStrings is not a translatable object`);
+          continue;
+        }
+
+        // All languages must have the same set of params.
+        for (const lang in templateObj) {
+          const template = templateObj[lang];
+          if (typeof template !== 'string') {
+            errorFunc(`${file}: '${key}' in '${trigger.id}' outputStrings for lang ${lang} is not a string`);
+            continue;
+          }
+
+          // Build params with a set for uniqueness, but store as an array later for ease of use.
+          const params = new Set();
+          template.replace(paramRegex, (fullMatch, key) => {
+            params.add(key);
+            return fullMatch;
+          });
+
+          // If this is not the first lang, validate it has the same params as previous languages.
+          if (key in outputStringsParams) {
+            // prevParams is an array, params is a set.
+            const prevParams = outputStringsParams[key];
+            let ok = false;
+            if (prevParams.length === params.size) {
+              ok = true;
+              for (const key of prevParams)
+                ok &= params.has(key);
+            }
+            if (!ok) {
+              errorFunc(`${file}: '${key}' in '${trigger.id}' outputStrings has inconsistent params among languages`);
+              continue;
+            }
+          }
+          outputStringsParams[key] = [...params];
+
+          // Verify that there's no dangling ${
+          if (/\${/.test(template.replace(paramRegex, ''))) {
+            errorFunc(`${file}: '${key}' in '${trigger.id}' outputStrings has a dangling \${`);
```

Comment:
I understand this and I feel it's concise, but I don't know how well it translates both across languages and to people that may be unfamiliar? Something like: "open ${ without matching closing }"?
Feel free to disregard if you think it's obvious.

Reply:
Nope, thanks! This is a good suggestion.